internal/sdkprovider/kafkatopicrepository/repository_test.go

```diff
@@ -0,0 +1,285 @@
+			v2ListCalled:    5,
+			v2ListBatchSize: 2,
+		},
+		{
+			name: "v1List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("bla bla bla")},
+			},
+			v1ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    0,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List bla bla bla error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: bla bla bla")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       fmt.Errorf("bla bla bla"),
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+		{
+			name: "v2List 404 error",
+			requests: []request{
+				{project: "a", service: "b", topic: "c"},
+			},
+			responses: []response{
+				{err: fmt.Errorf("topic read error: All attempts fail:\n#1: topic list has changed")},
+			},
+			storage: map[string]*aiven.KafkaListTopic{
+				"a/b/c": {TopicName: "c"},
+			},
+			v2ListErr:       aiven.Error{Status: 404},
+			v1ListCalled:    1,
+			v2ListCalled:    1,
+			v2ListBatchSize: defaultV2ListBatchSize,
+		},
+	}
+
+	for _, opt := range cases {
+		t.Run(opt.name, func(t *testing.T) {
+			client := &fakeTopicClient{
+				storage:   opt.storage,
+				v1ListErr: opt.v1ListErr,
+				v2ListErr: opt.v2ListErr,
+			}
+
+			ctx := context.Background()
+			rep := newRepository(client)
+			rep.workerCallInterval = time.Millisecond
+			rep.v2ListBatchSize = opt.v2ListBatchSize
+
+			// We must run all calls in parallel
+			// and then call the worker
+			var wg sync.WaitGroup
+			for i := range opt.requests {
+				wg.Add(1)
+				go func(i int) {
+					defer wg.Done()
+					topic, err := rep.Read(ctx, opt.requests[i].project, opt.requests[i].service, opt.requests[i].topic)
+					assert.Equal(t, opt.responses[i].topic, topic)
+					assert.Equal(t, opt.responses[i].err, err)
+				}(i)
+			}
+
+			go rep.worker()
+			wg.Wait()
+
+			assert.Equal(t, opt.v1ListCalled, client.v1ListCalled)
+			assert.Equal(t, opt.v2ListCalled, client.v2ListCalled)
+		})
+	}
+}
+
+var _ topicsClient = &fakeTopicClient{}
+
+type fakeTopicClient struct {
+	storage   map[string]*aiven.KafkaListTopic // key project/service/topic
+	createErr error
+	deleteErr error
+	v1ListErr error
+	v2ListErr error
+	// counters
+	createCalled int32
+	deleteCalled int32
+	v1ListCalled int32
+	v2ListCalled int32
+}
+
+func (f *fakeTopicClient) Create(context.Context, string, string, aiven.CreateKafkaTopicRequest) error {
+	time.Sleep(time.Millisecond * 100) // we need some lag to simulate races
+	atomic.AddInt32(&f.createCalled, 1)
+	return f.createErr
+}
+
+func (f *fakeTopicClient) Update(context.Context, string, string, string, aiven.UpdateKafkaTopicRequest) error {
+	panic("implement me")
```

Comment:
consider returning a not implemented error instead of causing a panic

Reply:
That's for panic is made. Program cannot proceed.